pywemo/ouimeaux_device/api/rules.py

```diff
@@ -0,0 +1,695 @@
+        sql = (
+            f"INSERT OR REPLACE INTO {self.TABLE_NAME} ({column_list_str}) "
+            f"VALUES ({value_placeholders})"
+        )
+        try:
+            cursor.execute(sql, values)
+        except sqlite3.Error:
+            LOG.exception("Query failed: %s %s", sql, values)
+            raise
+        try:
+            pk_name = self.primary_key_name()
+        except RuntimeError:
+            pass
+        else:
+            pk_type = self.FIELDS[pk_name]
+            if pk_type.auto_increment and not hasattr(self, pk_name):
+                setattr(self, pk_name, cursor.lastrowid)
+
+    def remove_from_db(self, cursor):
+        """Remove the Row from the sqlite database."""
+        pk_name = self.primary_key_name()
+        pk_value = self.primary_key_value()
+        sql = f"DELETE FROM {self.TABLE_NAME} WHERE {pk_name}=?"
+        try:
+            cursor.execute(sql, (pk_value,))
+        except sqlite3.Error:
+            LOG.exception("Query failed: %s %s", sql, pk_value)
+            raise
+
+
+class SQLType:
+    """Base class for custom sqlite schema types."""
+
+    TYPE_MAP = {
+        int: "INTEGER",
+        float: "REAL",
+        str: "TEXT",
+    }
+
+    def __init__(self, type_constructor, *, sql_type=None, not_null=False):
+        """Create a SQLType.
+
+        Args:
+            type_constructor: Callable(value) that will receive the value from
+              the database cursor and convert it to a Python type.
+            sql_type: The sqlite field type.
+            not_null: Whether or not the sqlite field can be null.
+        """
+        self.type_constructor = type_constructor
+        self._sql_type = (
+            sql_type
+            if sql_type is not None
+            else SQLType.TYPE_MAP[type_constructor]
+        )
+        self.not_null = not_null
+
+    def __call__(self, value):
+        """Convert the sqlite row value to a Python type."""
+        return self.type_constructor(value)
+
+    @property
+    def sql_type(self):
+        """Return the sqlite type name for this type."""
+        if self.not_null:
+            return f"{self._sql_type} NOT NULL".lstrip()
+        return self._sql_type.lstrip()
+
+
+class PrimaryKey(SQLType):
+    """Class used to indicate the primary key field for a Row."""
+
+    def __init__(self, *args, auto_increment=False, **kwargs):
+        """Create a PrimaryKey instance."""
+        super().__init__(*args, **kwargs)
+        self.auto_increment = auto_increment
+
+    @property
+    def sql_type(self):
+        """Return the sqlite type name for this type."""
+        value = super().sql_type
+        value = f"{value} PRIMARY KEY"
+        if self.auto_increment:
+            value = f"{value} AUTOINCREMENT"
+        return value.lstrip()
+
+
+class RulesRow(DatabaseRow):
+    """Row schema for the RULES table."""
+
+    TABLE_NAME = "RULES"
+    FIELDS = {
+        "RuleID": PrimaryKey(int, sql_type=""),
```

Comment:
Is `int` the right type for this `RuleID`?  From [a comment in the reference issue](https://github.com/pavoni/pywemo/issues/61#issuecomment-748693894), the field type is `RuleID PRIMARY KEY` and when inserted it uses `'501'`, as a string?

Reply:
Nice catch. This is interesting. RuleID seems as-if it is meant to match with the other RuleID values in the other tables (which are all INTEGER values in the database). From a Python standpoint, it's important that these all be the same type so comparisons work correctly (RulesRow.RuleID == RuleDevicesRow.RuleID should match). The database seems to be handling the INTEGER/TEXT type juggling just fine. However if there was ever a non-integer TEXT value in the database that would cause issues for pywemo.

I think it's okay to keep as an int within pywemo. The alternative could be to convert all the RuleID values to a str type in pywemo and keep their sql_type as INTEGER. What do you think?

Reply:
https://www.sqlite.org/faq.html#q3

> Data of any type can (usually) be inserted into any column. You can put arbitrary length strings into integer columns, floating point numbers in boolean columns, or dates in character columns.

Ooof! I didn't realize this. That explains why it is okay to store `NOSYNC` in an INTEGER column. If the types of the fields are not guaranteed, then I'm feeling less confident that the ORM library is a good idea now. If we're already seeing cases where string values are in integer fields, it stands to reason that other integer fields could become strings in the future. That could end up breaking pywemo due to the type juggling in the ORM library and increasing the cost to maintain that code.

I'm starting to think the right thing to do is to not implement that ORM layer and do field comparisons within the database; either with JOIN statements or with passing fetched values back into the WHERE clauses. That'll avoid future maintenance issues due to the wrong type being assumed in the ORM Row classes.

What do you think?

Reply:
Huh.  I'm not particularly experienced with SQL, so this certainly surprises me!  It does also indicate that using the integer PK though is 1 case where it does matter!  Weird to me that Wemo didn't decide to use an integer for that PK indicated above.  And those docs also indicate that an attempt will be made to convert it if a type is provided....so if `'501'` was provided to an integer column, then it should be converted to `501`.

As for your idea of removing the ORM layer...I'm not sure that it would "fix" this case indicated here?  If we compared the string `'501'` with integer `501`, then it wouldn't match anyway?  I guess Belkin must be doing a conversion within their code to look it up?

I don't suspect that Belkin would be motivated to change the types of their data with any sort of frequency, so that seems pretty low risk to me.  As you found though, they can put whatever they want in the column anyway.  Long story short, I'm not sure exactly what this code would look like without the ORM layer and I'm probably not experienced enough with SQL to make the judgement on which way is better here.  I'd need to fall back on your judgement there.

Reply:
> I don't suspect that Belkin would be motivated to change the types of their data with any sort of frequency, so that seems pretty low risk to me.

I think you're right. Let's go with what we have now. We can revisit in the future if there turns out to be changes.